kratos/utilities/voxel_utilities.cpp

```diff
@@ -0,0 +1,307 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ariadna Cortes
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "voxel_utilities.h"
+
+namespace Kratos 
+{
+constexpr std::size_t VoxelUtilities::mNeighbours[4][2];
+
+/**********************************************************************************/
+/**********************************************************************************/
+
+double VoxelUtilities::NodesApproximation(const GeometryType& rVoxel) {
+    double volume = 0;
+    const auto& nodes = rVoxel.Points(); //PointsArrayType
+    const std::size_t n = nodes.size();
+    for (std::size_t i = 0; i < n; i++) {
+        if (nodes[i].GetSolutionStepValue(DISTANCE) > 0) {
+            volume+=(1.0/nodes.size()); 
+        } 
+    }
+    return volume;
+}
+
+/**********************************************************************************/
+/**********************************************************************************/
+
+double VoxelUtilities::EdgesPortionApproximation(
+    const GeometryType& rVoxel,  
+    const GeometryArrayType& rTriangles     
+) {
+    double volume = 0;
+    GeometryArrayType edges = rVoxel.GenerateEdges();
+    std::vector<double> distances;
+    const std::size_t n = edges.size();
+    for (std::size_t i = 0; i < n; i++) {
+        distances.push_back(0);
+        const auto& ends = edges[i].Points();
+
+        for (auto& r_triangle : rTriangles) {
+            array_1d<double,3> intersection;
+            int result = IntersectionUtilities::ComputeTriangleLineIntersection(r_triangle,ends[0],ends[1],intersection);
+            
+            if(result == 1) {
+                const double dist = norm_2(ends[0].Coordinates() - intersection);
+                distances.push_back(dist);
+            }  
+        } 
+        distances.push_back(norm_2(ends[0].Coordinates() - ends[1].Coordinates()));
+        std::sort(distances.begin(),distances.end());        
+        const double edge_portion = VoxelUtilities::EdgeFilledPortion(distances, ends);
+        volume += edge_portion/edges.size();  
+        distances.clear();              
+    }
+    return volume;
+}
+
+/**********************************************************************************/
+/**********************************************************************************/
+
+double VoxelUtilities::FaceArea(
+    const GeometryType& rFace,  
+    const GeometryArrayType& rTriangles     
+) {
+    double area = 0;
+    GeometryArrayType edges = rFace.GenerateEdges();
+    auto nodes = rFace.Points(); 
+    const double face_area = PointsArea(nodes);
+    std::vector<std::pair<double,double>> min_distance_to_node(edges.size(),{1,1}); 
```

Comment:
if your algorithm only works for quadrilateral, I would suggest to use an array of size 4. (And check if the face is quadrilateral)